quickwit/quickwit-search/src/collector.rs

```diff
@@ -293,25 +233,64 @@ enum AggregationSegmentCollectors {
 pub struct QuickwitSegmentCollector {
     num_hits: u64,
     split_id: String,
-    score_extractor: SortingFieldExtractorPair,
+    score_extractors: Vec<SortingFieldExtractorComponent>,
     // PartialHits in this heap don't contain a split_id yet.
     top_k_hits: TopK<PartialHit, PartialHitSortingKey, HitSortingMapper>,
     segment_ord: u32,
     timestamp_filter_opt: Option<TimestampFilter>,
     aggregation: Option<AggregationSegmentCollectors>,
+    search_after: Option<PartialHit>,
 }
 
 impl QuickwitSegmentCollector {
     #[inline]
     fn collect_top_k(&mut self, doc_id: DocId, score: Score) {
-        let (sort_value, sort_value2) =
-            self.score_extractor.extract_typed_sort_value(doc_id, score);
+        let sort_values: Vec<_> = self
+            .score_extractors
+            .iter()
+            .map(|extractor| extractor.extract_typed_sort_value_opt(doc_id, score))
+            .collect();
+
+        if let Some(search_after) = &self.search_after {
+            let search_after_values = [
+                search_after.sort_value.and_then(|v| v.sort_value),
+                search_after.sort_value2.and_then(|v| v.sort_value),
+            ];
+            let mut cmp_result = compare_fields(
+                &self.top_k_hits.sort_key_mapper.orders,
+                &sort_values,
+                &search_after_values,
+            );
+            if !search_after.split_id.is_empty() {
+                // TODO actually it's not first, it should be what's in _shard_doc then first then
+                // default
+                let order = self
+                    .top_k_hits
+                    .sort_key_mapper
+                    .orders
+                    .first()
+                    .unwrap_or(&SortOrder::Desc);
+                cmp_result = cmp_result
+                    .then_with(|| order.compare(&self.split_id, &search_after.split_id))
```

Comment:
Can we compute that when we build the `SegmentCollector`? Comparing strings in collect is extremely expensive.